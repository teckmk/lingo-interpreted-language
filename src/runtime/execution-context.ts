export type StackFrame = {
  functionName: string
  filename: string
  position: {
    start: { line: number; column: number }
    end: { line: number; column: number }
  }
}

export class ExecutionContext {
  private callStack: StackFrame[]

  constructor() {
    this.callStack = []
  }

  public pushCallStack(frame: StackFrame) {
    this.callStack.push(frame)
  }

  public popCallStack() {
    this.callStack.pop()
  }

  public getCallStack(): StackFrame[] {
    return this.callStack.reverse()
  }

  static formatStackTrace(stack: StackFrame[], maxFrames = 10): string {
    const trimmedStack = stack.slice(0, maxFrames) // Limit frames to avoid excessive output
    let stackTrace = "Stack Trace:\n"
    for (const frame of trimmedStack) {
      // const shortFile = frame.filename.replace(/^.*[\\/]/, "") // Extract filename from path
      const shortFile = frame.filename
      stackTrace += `  at ${frame.functionName || "<anonymous>"} (${shortFile}:${frame.position.start.line}:${frame.position.start.column})\n`
    }

    if (stack.length > maxFrames) {
      stackTrace += `  ... (${stack.length - maxFrames} more frames)\n`
    }

    return stackTrace
  }
}
