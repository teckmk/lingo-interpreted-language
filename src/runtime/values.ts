import { Stmt } from "../frontend/2-ast"
import Environment from "./environment"

export type ValueType =
  | "null"
  | "number"
  | "string"
  | "boolean"
  | "object"
  | "nativefn"
  | "function"
  | "return"
  | "paramter"
  | "conditional"
  | "whileloop"
  | "array"
  | "dynamic"
  | "docs"

export interface RuntimeVal {
  type: ValueType
  returned: boolean
}

export interface NullVal extends RuntimeVal {
  type: "null"
  value: null
}
export interface BooleanVal extends RuntimeVal {
  type: "boolean"
  value: boolean
}

export interface NumberVal extends RuntimeVal {
  type: "number"
  value: number
}

export interface StringVal extends RuntimeVal {
  type: "string"
  value: string
}

export interface ObjectVal extends RuntimeVal {
  type: "object"
  properties: Map<string, RuntimeVal>
}

export type FunctionCall = (args: RuntimeVal[], env: Environment) => RuntimeVal

export interface NativeFnVal extends RuntimeVal {
  type: "nativefn"
  call: FunctionCall
}

export interface ParamVal extends RuntimeVal {
  type: "paramter"
  name: string
  valueType?: ValueType
  default: RuntimeVal // to assign a default value
}

export interface FunctionVal extends RuntimeVal {
  type: "function"
  name: string
  parameters: ParamVal[]
  declarationEnv: Environment
  body: Stmt[]
}

export interface ReturnVal extends RuntimeVal {
  type: "return"
  value: RuntimeVal
}

export interface ConditionalVal extends RuntimeVal {
  type: "conditional"
  check: RuntimeVal
  body: Stmt[]
  childChecks?: ConditionalVal[]
  else?: Stmt[]
}

export interface WhileLoopVal extends RuntimeVal {
  type: "whileloop"
  check: RuntimeVal
  body: Stmt[]
}

export interface ArrayVal extends RuntimeVal {
  type: "array"
  elements: RuntimeVal[]
}

export interface DocCommentVal extends RuntimeVal {
  type: "docs"
}
