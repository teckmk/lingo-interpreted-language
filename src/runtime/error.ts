import { ExecutionContext } from "./execution-context"

export class RuntimeError extends Error {
  context?: ExecutionContext

  constructor(context: ExecutionContext, message: string) {
    super(message)
    this.name = "RuntimeError"
    this.context = context
  }
}
