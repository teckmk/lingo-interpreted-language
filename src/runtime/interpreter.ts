import Environment from "./environment"
import { NumberVal, RuntimeVal, StringVal } from "./values"
import {
  ArrayLiteral,
  AssignmentExpr,
  BinaryExpr,
  CallExpr,
  DocComment,
  FunctionDeclaration,
  Identifier,
  IfElseStatement,
  MemberExpr,
  MultiVarDeclaration,
  NumericLiteral,
  ObjectLiteral,
  Program,
  ReturnStatement,
  Stmt,
  StringLiteral,
  VarDeclaration,
  WhileStatement,
} from "../frontend/2-ast"

import {
  eval_fn_declaration,
  eval_if_else_statement,
  eval_multi_var_declaration,
  eval_program,
  eval_return_statement,
  eval_var_declaration,
  eval_while_statement,
} from "./eval/statements"
import {
  eval_array_expr,
  eval_assignment,
  eval_binary_expr,
  eval_call_expr,
  eval_comment_expr,
  eval_identifier,
  eval_member_expr,
  eval_object_expr,
  eval_string_literal,
} from "./eval/expression"

export function evaluate(astNode: Stmt, env: Environment): RuntimeVal {
  switch (astNode.kind) {
    case "NumericLiteral":
      return {
        value: (astNode as NumericLiteral).value,
        type: "number",
      } as NumberVal
    case "StringLiteral":
      return eval_string_literal(astNode as StringLiteral, env) as StringVal
    case "BinaryExpr":
      return eval_binary_expr(astNode as BinaryExpr, env)
    case "AssignmentExpr":
      return eval_assignment(astNode as AssignmentExpr, env)
    case "Identifier":
      return eval_identifier(astNode as Identifier, env)
    case "ObjectLiteral":
      return eval_object_expr(astNode as ObjectLiteral, env)
    case "ArrayLiteral":
      return eval_array_expr(astNode as ArrayLiteral, env)
    case "CallExpr":
      return eval_call_expr(astNode as CallExpr, env)
    case "Program":
      return eval_program(astNode as Program, env)
    case "VarDeclaration":
      return eval_var_declaration(astNode as VarDeclaration, env)
    case "MultiVarDeclaration":
      return eval_multi_var_declaration(astNode as MultiVarDeclaration, env)
    case "FunctionDeclaration":
      return eval_fn_declaration(astNode as FunctionDeclaration, env)
    case "ReturnStatement":
      return eval_return_statement(astNode as ReturnStatement, env)
    case "IfElseStatement":
      return eval_if_else_statement(astNode as IfElseStatement, env)
    case "WhileStatement":
      return eval_while_statement(astNode as WhileStatement, env)
    case "MemberExpr":
      return eval_member_expr(astNode as MemberExpr, env)
    case "DocComment":
      return eval_comment_expr(astNode as DocComment, env)
    default:
      console.error("This AST can't be interpreted! For now atleast!", astNode)
      process.exit(0)
  }
}
