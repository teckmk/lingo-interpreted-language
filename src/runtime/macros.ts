import { BooleanVal, FunctionCall, NativeFnVal, NullVal, NumberVal, PlaceholderVal } from "./values"
import { PrimitiveTypeVal } from "./values.types"

export function MK_NUMBER(n = 0) {
  return { type: "number", value: n } as NumberVal
}

export function MK_NULL() {
  return { type: "null", value: null } as NullVal
}

export function MK_PLACEHOLDER() {
  return { type: "placeholder" } as PlaceholderVal
}

export function MK_BOOL(b: boolean) {
  return { type: "boolean", value: b } as BooleanVal
}

export function MK_NATIVE_FN(call: FunctionCall) {
  return { type: "nativefn", call } as NativeFnVal
}

export function MK_PRIM_TYPE(name: "number" | "bool" | "string" | "dynamic"): PrimitiveTypeVal {
  return {
    type: "type",
    typeName: name,
    typeKind: "primitive",
    primitiveType: name,
    returned: false, // To satify TS
  }
}
