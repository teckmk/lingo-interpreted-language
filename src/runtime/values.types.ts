import { RuntimeVal } from "./values"

export interface TypeVal extends RuntimeVal {
  type: "type"
  typeKind: string
  typeName?: string
  isNominal?: boolean // Track if this is a nominal type (created with 'type' keyword)
}

export interface PrimitiveTypeVal extends TypeVal {
  typeKind: "primitive"
  primitiveType: "string" | "number" | "bool" | "dynamic"
}

export interface StructTypeVal extends TypeVal {
  typeKind: "struct"
  members: Record<string, TypeVal>
  optional: Record<string, boolean> // Track which members are optional
}

export interface AliasTypeVal extends TypeVal {
  typeKind: "alias"
  aliasTo: TypeVal
}

export interface TypeParameterVal extends TypeVal {
  type: "type"
  typeKind: "typeParameter"
  name: string
  constraint?: TypeVal // i.e. T: number
}

export interface GenericTypeVal extends TypeVal {
  typeKind: "generic"
  baseType: TypeVal
  parameters: TypeParameterVal[]
}

export interface UnionTypeVal extends TypeVal {
  typeKind: "union"
  unionTypes: TypeVal[]
}

export interface ArrayTypeVal extends TypeVal {
  typeKind: "array"
  elementType: TypeVal
}
