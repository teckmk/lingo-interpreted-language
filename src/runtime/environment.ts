import { VarModifier } from "../frontend/2-ast"
import { MK_BOOL, MK_NATIVE_FN, MK_NULL } from "./macros"
import {
  ArrayVal,
  RuntimeVal,
  BooleanVal,
  FunctionVal,
  NumberVal,
  StringVal,
  NullVal,
  ValueType,
  ReturnVal,
} from "./values"

export default class Environment {
  private parent?: Environment
  private variables: Map<string, RuntimeVal>
  private constants: Set<string>
  private finals: Set<string>

  constructor(parentEnv?: Environment) {
    this.parent = parentEnv
    this.variables = new Map()
    this.constants = new Set()
    this.finals = new Set()

    const global = Boolean(parentEnv) == false
    if (global) {
      // global variables
      this.declareVar("true", MK_BOOL(true), "constant")
      this.declareVar("false", MK_BOOL(false), "constant")
      this.declareVar("null", MK_NULL(), "constant")

      // global native functions
      this.declareVar(
        "print",
        MK_NATIVE_FN((args: RuntimeVal[], _: Environment) => {
          const getValue = (arg: RuntimeVal): any => {
            const argType = arg.type

            switch (argType) {
              case "string":
                return (arg as StringVal).value
              case "number":
                return (arg as NumberVal).value
              case "boolean":
                return (arg as BooleanVal).value
              case "null":
                return (arg as NullVal).value
              case "array":
                return (arg as ArrayVal).elements.map(getValue)
              case "return":
                return getValue((arg as ReturnVal).value)
              case "function": {
                const fn = arg as FunctionVal
                return `fn ${fn.name}(${fn.parameters
                  .map((p) => `${p.name}: ${p.valueType || "dynamic"}`)
                  .join()})`
              }

              default:
                return arg
            }
          }

          console.log(...args.map((arg) => getValue(arg)))

          return MK_NULL()
        }),
        "final"
      )

      this.declareVar(
        "length",
        MK_NATIVE_FN((args: any[], _: Environment) => {
          const arr = args[0] as ArrayVal
          if (arr.type !== "array") throw new Error(`Cannot get length of type '${arr.type}'`)
          return { type: "number", value: arr.elements.length, returned: false }
        }),
        "final"
      )
    }
  }

  private assertType(varType: ValueType, value: RuntimeVal) {
    let valueWithType = value

    if (value.type == "dynamic") {
      valueWithType = { ...value, type: "dynamic" }
    } else if (value.type != varType) {
      throw new Error(`Can't assign a value of type ${value.type} to a variable of type ${varType}`)
    }

    return valueWithType
  }

  public declareVar(
    varname: string,
    value: RuntimeVal,
    modifier: VarModifier,
    type?: ValueType
  ): RuntimeVal {
    if (this.variables.has(varname))
      throw new Error(`Cannot declare variable ${varname}. As it already is defined.`)

    switch (modifier) {
      case "final":
        this.finals.add(varname)
        break
      case "constant":
        this.constants.add(varname)
        break
    }

    let valueWithType = value
    if (type) valueWithType = this.assertType(type, value)
    this.variables.set(varname, valueWithType)

    return valueWithType
  }

  public assignVar(varname: string, value: RuntimeVal): RuntimeVal {
    const env = this.resolve(varname)

    // Can't assign to a constant or final variable
    if (this.finals.has(varname)) throw new Error(`Cannot assign to a final variable "${varname}"`)
    else if (this.constants.has(varname))
      throw new Error(`Cannot assign to a constant variable "${varname}"`)

    const prevVal = env.lookupVar(varname)

    const valueWithType = this.assertType(prevVal.type, value)

    env.variables.set(varname, valueWithType)

    return valueWithType
  }

  public resolve(varname: string): Environment {
    if (this.variables.has(varname)) return this

    if (this.parent == undefined) throw new Error(`Cannot resolve ${varname} as it does not exist.`)

    return this.parent.resolve(varname)
  }

  public lookupVar(varname: string): RuntimeVal {
    const env = this.resolve(varname)
    return env.variables.get(varname) as RuntimeVal
  }
}
