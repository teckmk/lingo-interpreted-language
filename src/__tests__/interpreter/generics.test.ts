import { interpret } from "../../runtime/interpreter";

describe("Interpreter - Generics", () => {
  it("should interpret generics", () => {
    const code = `type Box<T> struct { \n      value: T \n    }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toMatchSnapshot();
  });
  it("should interpret generics with multiple type parameters", () => {
    const code = `type Pair<T, U> struct { \n      first: T \n      second: U \n    }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toMatchSnapshot();
  });
  it("should interpret generics with type parameters with constraints", () => {
    const code = `type Box<T: number> struct { \n      value: T \n    }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toMatchSnapshot();
  });
  it("should interpret generics with type parameters with constraints and multiple constraints", () => {
    const code = `type Box<T: number | string> struct { \n      value: T \n    }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toMatchSnapshot();
  });
  it("should interpret generics with type parameters with constraints and multiple constraints and multiple type parameters", () => {
    const code = `type Pair<T: number | string, U: number> struct { \n      first: T \n      second: U \n    }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toMatchSnapshot();
  });
});

describe("Interpreter - Generics - Instantiation", () => {
  it("should interpret generic type instantiation", () => {
    const code = `type Box<T> struct { \n      value: T \n    }
      let box = Box<number> { value: 10 }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toEqual({
      type: "object",
      returned: false,
      instanceOf: "Box",
      properties: new Map([["value", { type: "number", value: 10 }]]),
    });
  });

  it("should interpret generic type instantiation with multiple type parameters", () => {
    const code = `type Pair<T, U> struct { \n      first: T \n      second: U \n    }
      let pair = Pair<number, string> { first: 10, second: "hello" }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toEqual({
      type: "object",
      returned: false,
      instanceOf: "Pair",
      properties: new Map([
        ["first", { type: "number", value: 10 }],
        ["second", { type: "string", value: "hello" }],
      ]),
    });
  });

  it("should interpret generic type instantiation with type parameters with constraints", () => {
    const code = `type Box<T: number> struct { \n      value: T \n    }
      let box = Box<number> { value: 10 }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toEqual({
      type: "object",
      returned: false,
      instanceOf: "Box",
      properties: new Map([["value", { type: "number", value: 10 }]]),
    });
  });

  it("should interpret generic type instantiation with type parameters with constraints and multiple constraints", () => {
    const code = `type Box<T: number | string> struct { \n      value: T \n    }
      let box = Box<number> { value: 10 }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toEqual({
      type: "object",
      returned: false,
      instanceOf: "Box",
      properties: new Map([["value", { type: "number", value: 10 }]]),
    });
  });

  it("should interpret generic type instantiation with type parameters with constraints and multiple constraints and multiple type parameters", () => {
    const code = `type Pair<T: number | string, U: number> struct { \n      first: T \n      second: U \n    }
      let pair = Pair<number, number> { first: 10, second: 20 }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toEqual({
      type: "object",
      returned: false,
      instanceOf: "Pair",
      properties: new Map([
        ["first", { type: "number", value: 10 }],
        ["second", { type: "number", value: 20 }],
      ]),
    });
  });

  it("should throw error for generic type instantiation with wrong type parameters", () => {
    const code = `type Box<T: number> struct { \n      value: T \n    }
        let box = Box<string> { value: "hello" }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toBe("Type argument string does not satisfy constraint number of type parameter T\n");
  });

  it("should throw error for generic union type instantiation with wrong type parameters", () => {
    const code = `type Pair<T: number | string, U: number> struct { \n      first: T \n      second: U \n    }
      let pair = Pair<bool, number> { first: 10, second: 20 }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toEqual("Type argument bool does not satisfy constraint number | string of type parameter T\n");
  });
});
