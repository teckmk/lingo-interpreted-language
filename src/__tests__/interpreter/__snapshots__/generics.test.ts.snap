// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interpreter - Generics should interpret generics 1`] = `
{
  "baseType": {
    "members": {
      "value": {
        "aliasTo": {
          "primitiveType": "T",
          "type": "type",
          "typeKind": "primitive",
        },
        "type": "type",
        "typeKind": "alias",
        "typeName": "T",
      },
    },
    "optional": {
      "value": false,
    },
    "type": "type",
    "typeKind": "struct",
    "typeName": undefined,
  },
  "parameters": [
    {
      "constraint": undefined,
      "name": "T",
      "returned": false,
      "type": "type",
      "typeKind": "typeParameter",
    },
  ],
  "returned": false,
  "type": "type",
  "typeKind": "generic",
  "typeName": "Box",
}
`;

exports[`Interpreter - Generics should interpret generics with multiple type parameters 1`] = `
{
  "baseType": {
    "members": {
      "first": {
        "aliasTo": {
          "primitiveType": "T",
          "type": "type",
          "typeKind": "primitive",
        },
        "type": "type",
        "typeKind": "alias",
        "typeName": "T",
      },
      "second": {
        "aliasTo": {
          "primitiveType": "U",
          "type": "type",
          "typeKind": "primitive",
        },
        "type": "type",
        "typeKind": "alias",
        "typeName": "U",
      },
    },
    "optional": {
      "first": false,
      "second": false,
    },
    "type": "type",
    "typeKind": "struct",
    "typeName": undefined,
  },
  "parameters": [
    {
      "constraint": undefined,
      "name": "T",
      "returned": false,
      "type": "type",
      "typeKind": "typeParameter",
    },
    {
      "constraint": undefined,
      "name": "U",
      "returned": false,
      "type": "type",
      "typeKind": "typeParameter",
    },
  ],
  "returned": false,
  "type": "type",
  "typeKind": "generic",
  "typeName": "Pair",
}
`;

exports[`Interpreter - Generics should interpret generics with type parameters with constraints 1`] = `
{
  "baseType": {
    "members": {
      "value": {
        "aliasTo": {
          "primitiveType": "T",
          "type": "type",
          "typeKind": "primitive",
        },
        "type": "type",
        "typeKind": "alias",
        "typeName": "T",
      },
    },
    "optional": {
      "value": false,
    },
    "type": "type",
    "typeKind": "struct",
    "typeName": undefined,
  },
  "parameters": [
    {
      "constraint": {
        "primitiveType": "number",
        "type": "type",
        "typeKind": "primitive",
      },
      "name": "T",
      "returned": false,
      "type": "type",
      "typeKind": "typeParameter",
    },
  ],
  "returned": false,
  "type": "type",
  "typeKind": "generic",
  "typeName": "Box",
}
`;

exports[`Interpreter - Generics should interpret generics with type parameters with constraints and multiple constraints 1`] = `
{
  "baseType": {
    "members": {
      "value": {
        "aliasTo": {
          "primitiveType": "T",
          "type": "type",
          "typeKind": "primitive",
        },
        "type": "type",
        "typeKind": "alias",
        "typeName": "T",
      },
    },
    "optional": {
      "value": false,
    },
    "type": "type",
    "typeKind": "struct",
    "typeName": undefined,
  },
  "parameters": [
    {
      "constraint": {
        "type": "type",
        "typeKind": "union",
        "unionTypes": [
          {
            "primitiveType": "number",
            "type": "type",
            "typeKind": "primitive",
          },
          {
            "primitiveType": "string",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "name": "T",
      "returned": false,
      "type": "type",
      "typeKind": "typeParameter",
    },
  ],
  "returned": false,
  "type": "type",
  "typeKind": "generic",
  "typeName": "Box",
}
`;

exports[`Interpreter - Generics should interpret generics with type parameters with constraints and multiple constraints and multiple type parameters 1`] = `
{
  "baseType": {
    "members": {
      "first": {
        "aliasTo": {
          "primitiveType": "T",
          "type": "type",
          "typeKind": "primitive",
        },
        "type": "type",
        "typeKind": "alias",
        "typeName": "T",
      },
      "second": {
        "aliasTo": {
          "primitiveType": "U",
          "type": "type",
          "typeKind": "primitive",
        },
        "type": "type",
        "typeKind": "alias",
        "typeName": "U",
      },
    },
    "optional": {
      "first": false,
      "second": false,
    },
    "type": "type",
    "typeKind": "struct",
    "typeName": undefined,
  },
  "parameters": [
    {
      "constraint": {
        "type": "type",
        "typeKind": "union",
        "unionTypes": [
          {
            "primitiveType": "number",
            "type": "type",
            "typeKind": "primitive",
          },
          {
            "primitiveType": "string",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "name": "T",
      "returned": false,
      "type": "type",
      "typeKind": "typeParameter",
    },
    {
      "constraint": {
        "primitiveType": "number",
        "type": "type",
        "typeKind": "primitive",
      },
      "name": "U",
      "returned": false,
      "type": "type",
      "typeKind": "typeParameter",
    },
  ],
  "returned": false,
  "type": "type",
  "typeKind": "generic",
  "typeName": "Pair",
}
`;
